File: src/mars_uav_sizing/section4/derived_requirements.py
"""
Derived Requirements Calculations
==================================

Implements calculations from manuscript Section 4.12 (Derived Requirements).

Equations:
    @eq:stall-speed        - V_stall = sqrt(2×(W/S)/(ρ×C_L,max))
    @eq:wing-loading-limit - (W/S)_max = 0.5×ρ×V_min²×C_L,max
    @eq:reynolds-cruise    - Re = ρ×V×c/μ
    @eq:reynolds-chord     - c = Re×μ/(ρ×V)
    @eq:wing-area-from-ar  - S = c²×AR
    @eq:mach-cruise        - M = V/a

Reference:
    - Manuscript: sections_en/04_12_derived-requirements.md

Last Updated: 2025-12-29
"""

import math
from datetime import datetime
from typing import Dict, Any

from ..config import get_param


def stall_speed(wing_loading: float = None) -> float:
    """
    Calculate stall speed from wing loading.

    Implements @eq:stall-speed from §4.12:
        V_stall = sqrt(2 × (W/S) / (ρ × C_L,max))

    Parameters
    ----------
    wing_loading : float, optional
        Wing loading in N/m² (default: from Reynolds-based derivation)

    Returns
    -------
    float
        Stall speed in m/s
    """
    if wing_loading is None:
        # Use Reynolds-based wing loading as default
        wing_loading = wing_loading_from_reynolds()

    rho = get_param('environment.arcadia_planitia.density_kg_m3')
    cl_max = get_param('aerodynamic.airfoil.cl_max')

    return math.sqrt(2 * wing_loading / (rho * cl_max))


def minimum_flight_speed(wing_loading: float = None) -> float:
    """
    Calculate minimum flight speed with safety margin.
    
    V_min = factor × V_stall
    
    Parameters
    ----------
    wing_loading : float, optional
        Wing loading in N/m²
        
    Returns
    -------
    float
        Minimum safe flight speed in m/s
    """
    v_stall = stall_speed(wing_loading)
    factor = get_param('mission.velocity.v_min_factor')
    return factor * v_stall


def maximum_wing_loading(v_min: float = None) -> float:
    """
    Calculate maximum allowable wing loading from stall constraint.
    
    Implements @eq:wing-loading-limit from §4.12:
        (W/S)_max = 0.5 × ρ × V_min² × C_L,max
    
    Parameters
    ----------
    v_min : float, optional
        Minimum flight speed (default: from config)
        
    Returns
    -------
    float
        Maximum wing loading in N/m²
    """
    if v_min is None:
        v_stall = get_param('mission.velocity.v_stall_m_s')
        factor = get_param('mission.velocity.v_min_factor')
        v_min = v_stall * factor
    
    rho = get_param('environment.arcadia_planitia.density_kg_m3')
    cl_max = get_param('aerodynamic.airfoil.cl_max')
    
    return 0.5 * rho * v_min**2 * cl_max


def chord_from_reynolds(target_re: float = 60000) -> float:
    """
    Calculate required chord for target Reynolds number.

    Implements @eq:reynolds-chord from §4.12 (derived from @eq:reynolds-definition):
        c = Re × μ / (ρ × V)

    This is the fundamental equation linking Reynolds number target
    to required wing geometry in the Mars atmosphere.

    Parameters
    ----------
    target_re : float
        Target Reynolds number (default: 60,000 for E387 performance)

    Returns
    -------
    float
        Required mean chord in m
    """
    rho = get_param('environment.arcadia_planitia.density_kg_m3')
    v_cruise = get_param('mission.velocity.v_cruise_m_s')
    mu = get_param('environment.arcadia_planitia.viscosity_Pa_s')

    return target_re * mu / (rho * v_cruise)


def wing_area_from_reynolds(target_re: float = 60000, aspect_ratio: float = None) -> float:
    """
    Calculate required wing area from Reynolds number target.

    Implements @eq:wing-area-from-ar from §4.12:
        S = c̄² × AR

    where c̄ is derived from @eq:reynolds-chord.

    Parameters
    ----------
    target_re : float
        Target Reynolds number (default: 60,000)
    aspect_ratio : float, optional
        Aspect ratio (default: from config)

    Returns
    -------
    float
        Required wing area in m²
    """
    if aspect_ratio is None:
        aspect_ratio = get_param('aerodynamic.wing.aspect_ratio')

    chord = chord_from_reynolds(target_re)
    return chord**2 * aspect_ratio


def wing_loading_from_reynolds(mtow_kg: float = None, target_re: float = 60000) -> float:
    """
    Calculate wing loading consistent with Reynolds number target.

    Implements the relationship between Re, chord, and wing loading
    from @eq:reynolds-chord and wing geometry:
        W/S = W / S = (m × g) / (c̄² × AR)

    Parameters
    ----------
    mtow_kg : float, optional
        Maximum takeoff weight in kg (default: from config)
    target_re : float
        Target Reynolds number (default: 60,000)

    Returns
    -------
    float
        Wing loading in N/m²
    """
    if mtow_kg is None:
        mtow_kg = get_param('mission.mass.mtow_kg')

    g = get_param('physical.mars.g')
    weight_n = mtow_kg * g

    wing_area = wing_area_from_reynolds(target_re)
    return weight_n / wing_area


def cruise_reynolds(chord: float = None) -> float:
    """
    Calculate Reynolds number at cruise conditions.

    Implements @eq:reynolds-cruise from §4.12:
        Re = ρ × V × c / μ

    Parameters
    ----------
    chord : float, optional
        Wing chord in m (default: from config mean_chord_m)

    Returns
    -------
    float
        Reynolds number
    """
    if chord is None:
        chord = get_param('aerodynamic.wing.mean_chord_m')

    rho = get_param('environment.arcadia_planitia.density_kg_m3')
    v_cruise = get_param('mission.velocity.v_cruise_m_s')
    mu = get_param('environment.arcadia_planitia.viscosity_Pa_s')

    return rho * v_cruise * chord / mu


def cruise_mach() -> float:
    """
    Calculate Mach number at cruise.
    
    Implements @eq:mach-cruise from §4.12:
        M = V / a
    
    Returns
    -------
    float
        Mach number
    """
    v_cruise = get_param('mission.velocity.v_cruise_m_s')
    a = get_param('environment.arcadia_planitia.speed_of_sound_m_s')
    return v_cruise / a


def wing_area_from_loading(mtow_kg: float = None, wing_loading: float = None) -> float:
    """
    Calculate wing area from MTOW and wing loading.

    S = W / (W/S) = (m × g) / (W/S)

    Parameters
    ----------
    mtow_kg : float, optional
        Maximum takeoff weight in kg (default: from config)
    wing_loading : float, optional
        Wing loading in N/m² (default: from stall constraint)

    Returns
    -------
    float
        Wing area in m²
    """
    if mtow_kg is None:
        mtow_kg = get_param('mission.mass.mtow_kg')
    if wing_loading is None:
        # Default: compute from stall constraint (V_min-based)
        wing_loading = maximum_wing_loading()

    g = get_param('physical.mars.g')
    weight_n = mtow_kg * g

    return weight_n / wing_loading


def wingspan_from_area(wing_area: float, aspect_ratio: float = None) -> float:
    """
    Calculate wingspan from wing area and aspect ratio.
    
    b = sqrt(AR × S)
    
    Parameters
    ----------
    wing_area : float
        Wing area in m²
    aspect_ratio : float, optional
        Aspect ratio (default: from config)
        
    Returns
    -------
    float
        Wingspan in m
    """
    if aspect_ratio is None:
        aspect_ratio = get_param('aerodynamic.wing.aspect_ratio')
    
    return math.sqrt(aspect_ratio * wing_area)


def mean_chord(wing_area: float, wingspan: float) -> float:
    """
    Calculate mean aerodynamic chord.
    
    c = S / b
    
    Parameters
    ----------
    wing_area : float
        Wing area in m²
    wingspan : float
        Wingspan in m
        
    Returns
    -------
    float
        Mean chord in m
    """
    return wing_area / wingspan


def derived_requirements_analysis() -> Dict[str, Any]:
    """
    Complete derived requirements analysis.

    Includes both Reynolds-based geometry derivation (from §4.12 cruise velocity)
    and stall constraint-based wing loading (from §5.2 matching chart).

    Returns
    -------
    dict
        All derived requirements including both derivation approaches
    """
    g = get_param('physical.mars.g')
    mtow_kg = get_param('mission.mass.mtow_kg')
    weight_n = mtow_kg * g
    ar = get_param('aerodynamic.wing.aspect_ratio')

    # Velocity requirements
    v_cruise = get_param('mission.velocity.v_cruise_m_s')
    v_min_factor = get_param('mission.velocity.v_min_factor')

    # ==========================================================================
    # REYNOLDS-BASED DERIVATION (from §4.12)
    # Target Re = 60,000 → chord → wing area → wing loading
    # ==========================================================================
    target_re = 60000
    chord_re = chord_from_reynolds(target_re)
    wing_area_re = wing_area_from_reynolds(target_re, ar)
    wingspan_re = wingspan_from_area(wing_area_re, ar)
    ws_re = wing_loading_from_reynolds(mtow_kg, target_re)
    v_stall_re = stall_speed(ws_re)
    v_min_re = v_stall_re * v_min_factor

    # ==========================================================================
    # STALL CONSTRAINT-BASED DERIVATION (from §5.2 matching chart)
    # V_stall (config) → V_min → W/S_max → wing area
    # ==========================================================================
    v_stall_config = get_param('mission.velocity.v_stall_m_s')
    v_min_stall = v_stall_config * v_min_factor
    ws_stall = maximum_wing_loading(v_min_stall)
    wing_area_stall = wing_area_from_loading(mtow_kg, ws_stall)
    wingspan_stall = wingspan_from_area(wing_area_stall, ar)
    chord_stall = mean_chord(wing_area_stall, wingspan_stall)
    re_stall = cruise_reynolds(chord_stall)

    # Mach number (same for both)
    M = cruise_mach()

    return {
        # Mass
        'mtow_kg': mtow_kg,
        'weight_n': weight_n,

        # Velocity
        'v_cruise_m_s': v_cruise,
        'v_min_factor': v_min_factor,

        # Reynolds-based derivation (§4.12)
        'target_reynolds': target_re,
        're_chord_m': chord_re,
        're_wing_area_m2': wing_area_re,
        're_wingspan_m': wingspan_re,
        're_wing_loading': ws_re,
        're_v_stall_m_s': v_stall_re,
        're_v_min_m_s': v_min_re,

        # Stall-based derivation (matching chart)
        'stall_v_stall_m_s': v_stall_config,
        'stall_v_min_m_s': v_min_stall,
        'stall_wing_loading': ws_stall,
        'stall_wing_area_m2': wing_area_stall,
        'stall_wingspan_m': wingspan_stall,
        'stall_chord_m': chord_stall,
        'stall_reynolds': re_stall,

        # Common
        'mach_cruise': M,
        'aspect_ratio': ar,
    }


def print_analysis():
    """Print formatted derived requirements analysis."""
    results = derived_requirements_analysis()
    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")

    print("=" * 80)
    print("DERIVED REQUIREMENTS (Section 4.12)")
    print("=" * 80)
    print(f"Computed: {timestamp}")
    print(f"Config:   All values from config/ YAML files")
    print()

    print("MASS AND WEIGHT")
    print("-" * 50)
    print(f"  MTOW:               {results['mtow_kg']:.2f} kg")
    print(f"  Weight (Mars):      {results['weight_n']:.2f} N")
    print(f"  Aspect ratio:       {results['aspect_ratio']}")
    print()

    print("VELOCITY REQUIREMENTS")
    print("-" * 50)
    print(f"  V_cruise:           {results['v_cruise_m_s']:.2f} m/s")
    print(f"  V_min factor:       {results['v_min_factor']}")
    print(f"  Mach number:        {results['mach_cruise']:.4f}")
    print()

    print("=" * 80)
    print("DERIVATION APPROACH 1: Reynolds-based (Section 4.12)")
    print("  Target Re = 60,000 -> chord -> wing area -> W/S -> V_stall")
    print("=" * 80)
    print(f"  Target Reynolds:    {results['target_reynolds']:.0f}")
    print(f"  Required chord:     {results['re_chord_m']:.4f} m")
    print(f"  Wing area:          {results['re_wing_area_m2']:.4f} m2")
    print(f"  Wingspan:           {results['re_wingspan_m']:.4f} m")
    print(f"  Wing loading:       {results['re_wing_loading']:.4f} N/m2")
    print(f"  V_stall:            {results['re_v_stall_m_s']:.4f} m/s")
    print(f"  V_min (1.2xV_stall):{results['re_v_min_m_s']:.4f} m/s")
    print()

    print("=" * 80)
    print("DERIVATION APPROACH 2: Stall-constrained (Matching chart)")
    print("  V_stall (config) -> V_min -> W/S_max -> wing area -> chord -> Re")
    print("=" * 80)
    print(f"  V_stall (config):   {results['stall_v_stall_m_s']:.4f} m/s")
    print(f"  V_min (1.2xV_stall):{results['stall_v_min_m_s']:.4f} m/s")
    print(f"  Wing loading max:   {results['stall_wing_loading']:.4f} N/m2")
    print(f"  Wing area:          {results['stall_wing_area_m2']:.4f} m2")
    print(f"  Wingspan:           {results['stall_wingspan_m']:.4f} m")
    print(f"  Mean chord:         {results['stall_chord_m']:.4f} m")
    print(f"  Achieved Reynolds:  {results['stall_reynolds']:.0f}")
    print()

    print("=" * 80)
    print("COMPARISON")
    print("=" * 80)
    re_ws = results['re_wing_loading']
    stall_ws = results['stall_wing_loading']
    print(f"  W/S (Reynolds):     {re_ws:.4f} N/m2")
    print(f"  W/S (Stall):        {stall_ws:.4f} N/m2")
    print(f"  Ratio:              {stall_ws/re_ws:.4f}")
    print()
    print("  Note: The stall-based derivation determines W/S for the matching")
    print("        chart. The Reynolds-based derivation shows the geometry")
    print("        required to achieve Re=60,000 at cruise velocity.")
    print("=" * 80)


if __name__ == "__main__":
    print_analysis()

File: src/matlab_scripts/+mars_uav/+section4/derived_requirements.m
classdef derived_requirements
    %DERIVED_REQUIREMENTS Requisiti derivati (Sezione 4.12).

    methods(Static)
        function v_stall = stall_speed(wing_loading)
            %STALL_SPEED Velocita di stallo da carico alare.
            if nargin < 1 || isempty(wing_loading)
                wing_loading = mars_uav.section4.derived_requirements.wing_loading_from_reynolds();
            end
            rho = mars_uav.config.get_param('environment.arcadia_planitia.density_kg_m3');
            cl_max = mars_uav.config.get_param('aerodynamic.airfoil.cl_max');
            v_stall = sqrt(2 * wing_loading / (rho * cl_max));
        end

        function v_min = minimum_flight_speed(wing_loading)
            %MINIMUM_FLIGHT_SPEED Velocita minima con fattore di margine.
            v_stall = mars_uav.section4.derived_requirements.stall_speed(wing_loading);
            factor = mars_uav.config.get_param('mission.velocity.v_min_factor');
            v_min = factor * v_stall;
        end

        function ws_max = maximum_wing_loading(v_min)
            %MAXIMUM_WING_LOADING Limite massimo W/S da stallo.
            if nargin < 1 || isempty(v_min)
                v_stall = mars_uav.config.get_param('mission.velocity.v_stall_m_s');
                factor = mars_uav.config.get_param('mission.velocity.v_min_factor');
                v_min = v_stall * factor;
            end
            rho = mars_uav.config.get_param('environment.arcadia_planitia.density_kg_m3');
            cl_max = mars_uav.config.get_param('aerodynamic.airfoil.cl_max');
            ws_max = 0.5 * rho * v_min^2 * cl_max;
        end

        function chord = chord_from_reynolds(target_re)
            %CHORD_FROM_REYNOLDS Corda media richiesta per Re target.
            if nargin < 1 || isempty(target_re)
                target_re = 60000;
            end
            rho = mars_uav.config.get_param('environment.arcadia_planitia.density_kg_m3');
            v_cruise = mars_uav.config.get_param('mission.velocity.v_cruise_m_s');
            mu = mars_uav.config.get_param('environment.arcadia_planitia.viscosity_Pa_s');
            chord = target_re * mu / (rho * v_cruise);
        end

        function wing_area = wing_area_from_reynolds(target_re, aspect_ratio)
            %WING_AREA_FROM_REYNOLDS Area alare da Re target e AR.
            if nargin < 1 || isempty(target_re)
                target_re = 60000;
            end
            if nargin < 2 || isempty(aspect_ratio)
                aspect_ratio = mars_uav.config.get_param('aerodynamic.wing.aspect_ratio');
            end
            chord = mars_uav.section4.derived_requirements.chord_from_reynolds(target_re);
            wing_area = chord^2 * aspect_ratio;
        end

        function ws = wing_loading_from_reynolds(mtow_kg, target_re)
            %WING_LOADING_FROM_REYNOLDS W/S coerente con Re target.
            if nargin < 1 || isempty(mtow_kg)
                mtow_kg = mars_uav.config.get_param('mission.mass.mtow_kg');
            end
            if nargin < 2 || isempty(target_re)
                target_re = 60000;
            end
            g = mars_uav.config.get_param('physical.mars.g');
            weight_n = mtow_kg * g;
            wing_area = mars_uav.section4.derived_requirements.wing_area_from_reynolds(target_re);
            ws = weight_n / wing_area;
        end

        function re = cruise_reynolds(chord)
            %CRUISE_REYNOLDS Numero di Reynolds in crociera.
            if nargin < 1 || isempty(chord)
                chord = mars_uav.config.get_param('aerodynamic.wing.mean_chord_m');
            end
            rho = mars_uav.config.get_param('environment.arcadia_planitia.density_kg_m3');
            v_cruise = mars_uav.config.get_param('mission.velocity.v_cruise_m_s');
            mu = mars_uav.config.get_param('environment.arcadia_planitia.viscosity_Pa_s');
            re = rho * v_cruise * chord / mu;
        end

        function mach = cruise_mach()
            %CRUISE_MACH Numero di Mach in crociera.
            v_cruise = mars_uav.config.get_param('mission.velocity.v_cruise_m_s');
            a = mars_uav.config.get_param('environment.arcadia_planitia.speed_of_sound_m_s');
            mach = v_cruise / a;
        end

        function wing_area = wing_area_from_loading(mtow_kg, wing_loading)
            %WING_AREA_FROM_LOADING Area alare da MTOW e W/S.
            if nargin < 1 || isempty(mtow_kg)
                mtow_kg = mars_uav.config.get_param('mission.mass.mtow_kg');
            end
            if nargin < 2 || isempty(wing_loading)
                wing_loading = mars_uav.section4.derived_requirements.maximum_wing_loading();
            end
            g = mars_uav.config.get_param('physical.mars.g');
            weight_n = mtow_kg * g;
            wing_area = weight_n / wing_loading;
        end

        function wingspan = wingspan_from_area(wing_area, aspect_ratio)
            %WINGSPAN_FROM_AREA Apertura alare da area e AR.
            if nargin < 2 || isempty(aspect_ratio)
                aspect_ratio = mars_uav.config.get_param('aerodynamic.wing.aspect_ratio');
            end
            wingspan = sqrt(aspect_ratio * wing_area);
        end

        function chord = mean_chord(wing_area, wingspan)
            %MEAN_CHORD Corda media aerodinamica.
            chord = wing_area / wingspan;
        end

        function results = derived_requirements_analysis()
            %DERIVED_REQUIREMENTS_ANALYSIS Analisi requisiti derivati.
            g = mars_uav.config.get_param('physical.mars.g');
            mtow_kg = mars_uav.config.get_param('mission.mass.mtow_kg');
            weight_n = mtow_kg * g;
            ar = mars_uav.config.get_param('aerodynamic.wing.aspect_ratio');

            v_cruise = mars_uav.config.get_param('mission.velocity.v_cruise_m_s');
            v_min_factor = mars_uav.config.get_param('mission.velocity.v_min_factor');

            target_re = 60000;
            chord_re = mars_uav.section4.derived_requirements.chord_from_reynolds(target_re);
            wing_area_re = mars_uav.section4.derived_requirements.wing_area_from_reynolds(target_re, ar);
            wingspan_re = mars_uav.section4.derived_requirements.wingspan_from_area(wing_area_re, ar);
            ws_re = mars_uav.section4.derived_requirements.wing_loading_from_reynolds(mtow_kg, target_re);
            v_stall_re = mars_uav.section4.derived_requirements.stall_speed(ws_re);
            v_min_re = v_stall_re * v_min_factor;

            v_stall_config = mars_uav.config.get_param('mission.velocity.v_stall_m_s');
            v_min_stall = v_stall_config * v_min_factor;
            ws_stall = mars_uav.section4.derived_requirements.maximum_wing_loading(v_min_stall);
            wing_area_stall = mars_uav.section4.derived_requirements.wing_area_from_loading(mtow_kg, ws_stall);
            wingspan_stall = mars_uav.section4.derived_requirements.wingspan_from_area(wing_area_stall, ar);
            chord_stall = mars_uav.section4.derived_requirements.mean_chord(wing_area_stall, wingspan_stall);
            re_stall = mars_uav.section4.derived_requirements.cruise_reynolds(chord_stall);

            mach = mars_uav.section4.derived_requirements.cruise_mach();

            results = struct( ...
                'mtow_kg', mtow_kg, ...
                'weight_n', weight_n, ...
                'v_cruise_m_s', v_cruise, ...
                'v_min_factor', v_min_factor, ...
                'target_reynolds', target_re, ...
                're_chord_m', chord_re, ...
                're_wing_area_m2', wing_area_re, ...
                're_wingspan_m', wingspan_re, ...
                're_wing_loading', ws_re, ...
                're_v_stall_m_s', v_stall_re, ...
                're_v_min_m_s', v_min_re, ...
                'stall_v_stall_m_s', v_stall_config, ...
                'stall_v_min_m_s', v_min_stall, ...
                'stall_wing_loading', ws_stall, ...
                'stall_wing_area_m2', wing_area_stall, ...
                'stall_wingspan_m', wingspan_stall, ...
                'stall_chord_m', chord_stall, ...
                'stall_reynolds', re_stall, ...
                'mach_cruise', mach, ...
                'aspect_ratio', ar ...
            );
        end

        function print_analysis()
            %PRINT_ANALYSIS Stampa analisi requisiti derivati.
            results = mars_uav.section4.derived_requirements.derived_requirements_analysis();
            timestamp = datestr(now, 'yyyy-mm-dd HH:MM:SS');
            fmt = @(value, decimals) mars_uav.core.format_number(value, decimals);

            fprintf('%s\n', repmat('=', 1, 80));
            fprintf('REQUISITI DERIVATI (Sezione 4.12)\n');
            fprintf('%s\n', repmat('=', 1, 80));
            fprintf('Calcolato: %s\n', timestamp);
            fprintf('Config:    Parametri da file YAML in config/\n\n');

            fprintf('MASSA E PESO\n');
            fprintf('%s\n', repmat('-', 1, 50));
            fprintf('  MTOW:              %s kg\n', fmt(results.mtow_kg, 2));
            fprintf('  Peso (Marte):      %s N\n', fmt(results.weight_n, 2));
            fprintf('  Rapporto d''aspetto: %s\n\n', fmt(results.aspect_ratio, 2));

            fprintf('REQUISITI VELOCITA\n');
            fprintf('%s\n', repmat('-', 1, 50));
            fprintf('  V_crociera:        %s m/s\n', fmt(results.v_cruise_m_s, 2));
            fprintf('  Fattore V_min:     %s\n', fmt(results.v_min_factor, 2));
            fprintf('  Mach crociera:     %s\n\n', fmt(results.mach_cruise, 4));

            fprintf('%s\n', repmat('=', 1, 80));
            fprintf('DERIVAZIONE 1: Re target (Sezione 4.12)\n');
            fprintf('%s\n', repmat('=', 1, 80));
            fprintf('  Re target:         %s\n', fmt(results.target_reynolds, 0));
            fprintf('  Corda richiesta:   %s m\n', fmt(results.re_chord_m, 4));
            fprintf('  Superficie alare:  %s m^2\n', fmt(results.re_wing_area_m2, 4));
            fprintf('  Apertura alare:    %s m\n', fmt(results.re_wingspan_m, 4));
            fprintf('  Carico alare:      %s N/m^2\n', fmt(results.re_wing_loading, 4));
            fprintf('  V_stallo:          %s m/s\n', fmt(results.re_v_stall_m_s, 4));
            fprintf('  V_min (1.2x):      %s m/s\n\n', fmt(results.re_v_min_m_s, 4));

            fprintf('%s\n', repmat('=', 1, 80));
            fprintf('DERIVAZIONE 2: Stallo (matching chart)\n');
            fprintf('%s\n', repmat('=', 1, 80));
            fprintf('  V_stallo (config): %s m/s\n', fmt(results.stall_v_stall_m_s, 4));
            fprintf('  V_min (1.2x):      %s m/s\n', fmt(results.stall_v_min_m_s, 4));
            fprintf('  Carico alare max:  %s N/m^2\n', fmt(results.stall_wing_loading, 4));
            fprintf('  Superficie alare:  %s m^2\n', fmt(results.stall_wing_area_m2, 4));
            fprintf('  Apertura alare:    %s m\n', fmt(results.stall_wingspan_m, 4));
            fprintf('  Corda media:       %s m\n', fmt(results.stall_chord_m, 4));
            fprintf('  Re raggiunto:      %s\n\n', fmt(results.stall_reynolds, 0));

            fprintf('%s\n', repmat('=', 1, 80));
            fprintf('CONFRONTO\n');
            fprintf('%s\n', repmat('=', 1, 80));
            fprintf('  W/S (Re):          %s N/m^2\n', fmt(results.re_wing_loading, 4));
            fprintf('  W/S (stallo):      %s N/m^2\n', fmt(results.stall_wing_loading, 4));
            fprintf('  Rapporto:          %s\n', fmt(results.stall_wing_loading / results.re_wing_loading, 4));
            fprintf('%s\n', repmat('=', 1, 80));
        end
    end
end
